### DISCUSSION TITLE
Would it be possible to make libvips fork safe?

### DISCUSSION THREAD

**Discussion by casperisfine** [libvips/libvips#3577] [Ideas]
Somewhat related to https://github.com/libvips/libvips/discussions/2077, but I'm making a new discussion for clarity.

### Context

`libvips` and its dependencies use threads internally, so when an application called into `libvips` and later `fork(2)`, all threads but the one that called `fork` die, and can potentially leave mutexes and such locked with no way to unlock them. In other words after a `fork(2)`, all bets are off.

The issue is that in languages with a GIL like Ruby and Python, forking is a very common way to achieve parallelism, so `libvips` lack of fork safety makes it challenging to use in these environments.

### Question

How hard would it be to make `libvips` fully fork safe? Are you open to contributions / sponsoring on that front?

In https://github.com/libvips/libvips/discussions/2077 you mentioned that `vips_shutdown()` could be improved so users could  call `vips_init()` again later?

have you explored that idea further, was there any major blocker?

**Comment by jcupitt** [comment]
Hi @casperisfine,

My understanding is that fork and threads are fundamentally incompatible. Suppose you have a mutex (for example) that's being held by a thread, and then another thread forks. Now the child process has a copy of the mutex memory area (is locked, held by process 12345, etc.), but no thread 12345. Everything will break horribly.

libvips tries to work around this by delaying threads for as long as possible -- the first thread is only created when the first pixel is processed. However, once you have processed some pixels, the threading system is built, and forking after that point is likely to end in disaster.

Making `vips_shutdown()` really zero out everything would be a lot of work, I think. It would be much better, if possible, to delay pixel processing until after fork. 

I think what we really need is a test case that replicates the pattern of forking and threading that's failing in these cases. If we have that, maybe we can find some other way around this issue. It would ideally be small, complete and self-contained, but I realize that can be very hard to do :(

(I hope I've not misunderstood the problem, please put me right if I have)

**Reply by casperisfine** [reply]
> My understanding is that fork and threads are fundamentally incompatible. 

Yes and no. You can't just `fork` willy nilly, you need to ensure mutexes and such are unlocked before the fork, which usually involve shutting down the threads, somewhat similar to what `vips_shutdown()` is doing.

`grpc`, a very much threaded library, was also a blocker for us, and it was made fork safe this way: https://github.com/grpc/grpc/pull/33430. You need to call a `prefork` function before forking, and then a `postfork` function after.

It's involved, but not impossible.

> It would be much better, if possible, to delay pixel processing until after fork.

The `reforking` concept I linked to doesn't allow for that. Very quickly the idea is:

  - boot the application in one "master" process
  - fork X workers processes
  - once 1 worker process has processed Y requests, fork it again to replace the "master"
  - refork all workers from that new master.

The goal being to drastically improve Copy-on-Write performance, hence drastically reduce memory usage.

So we can't delay `vips` usage.

> I think what we really need is a test case that replicates the pattern of forking and threading that's failing in these cases.

So I'm only starting to poke at the code base, but the idea I had was to basically suround every single tests with `vips_init()` and `vips_shutdown()`, and perhaps some assertion to ensure only the main thread is alive after `vips_shutdown()`.

Again, I have only very limited knowledge of `libvips`, but it would seem to me that this would be sufficient coverage to ensure that the library remains functional after repeated uses of `init` and `shutdown`, hence is fork safe.

**Reply by jcupitt** [reply]
Ah I see, that does sound a bit tricky.

```
john@banana ~/GIT/libvips/libvips (dzsave-direct) $ grep mutex_new */*.c | grep -v ">" | wc
     11      41     628
```

So I suppose those are the static mutexes in libvips. They'd each need a `_shutdown()` function, called from `vips_shutdown()`. There are some static threads too, they'd need something similar, and perhaps some static semaphores.

A harder problem could be `g_once()`:

https://docs.gtk.org/glib/func.once.html

```
john@banana ~/GIT/libvips/libvips (dzsave-direct) $ grep -i gonce */*.c | wc
     20     119    1053
```

You might need to find some way to make them all resettable. Or maybe not? I'm unsure, you'd probably need to read all those init functions and check.

**Reply by casperisfine** [reply]
> They'd each need a _shutdown() function,

Hum, would they? Assuming the threads using them are cleanly shutdown, the mutex should be left unlocked, no?

> A harder problem could be g_once():

TIL.

It does indeed seem like a large enterprise. I will reach for workaround for now, I think in the application I care about Vips is infrequently used, so I might be able to roll with refoking being disabled once it has been used. That will require a way to check if `vips_init()` has been called.

**Reply by casperisfine** [reply]
Hum, actually I'm a bit confused here. 

`ruby-vips` calls `vips_init` as soon as it's loaded: https://github.com/libvips/ruby-vips/blob/acbdf3e8f4084b930d02fd72e4efeba41ee86852/lib/vips.rb#L617-L621.

And I'm 100% certain loaded during boot in our app and continues to work after the initial fork. There must be more lazy initialization going on later.

**Reply by jcupitt** [reply]
Yes, the threads that libvips uses are created on demand, so mutexes stay in their inited-but-unused state until pixels are processed.

**Reply by casperisfine** [reply]
Ah, that makes sense.

I'll have to dig more into the library, but this is low priority for now, our usage in this app is limited, so I'll just use a workaround short term.

Thank you very much for your insights!

**Comment by jcupitt** [comment]
So this should work:

1. main() starts
2. vips_init()
3. fork()
4. vips_image_new_from_file()
5. vips_image_write_to_file()

But if you move the fork after 4., and esp. if it's after 5., everything will break.

(again, that's my understanding, corrections very welcome)

**Reply by SamSaffron** [reply]
note will not work if VIPS_MAX_THREADS, it must be unset ...

**Reply by kleisauke** [reply]
> note will not work if VIPS_MAX_THREADS, it must be unset ...

@SamSaffron The undocumented and unsupported `VIPS_MAX_THREADS` environment variable should not be used outside WebAssembly, see for example:
https://github.com/libvips/libvips/pull/3105#issuecomment-1289090668

And:
```console
$ vips black x.jpg 100 100
$ VIPS_MAX_THREADS=4 vips autorot x.jpg x2.jpg --vips-concurrency=3 --vips-progress
```

**Reply by SamSaffron** [reply]
thanks, another thing I discovered (when it comes to Linux forking).

Doing this in the parent appears to be safe, and can knock off ~4ms of warmup from the child only leaving the thread creation at 0.5ms which is cheap. 

```
VipsOperation *black = vips_operation_new("black");
VipsOperation *avg = vips_operation_new("avg");
g_object_unref(black);
g_object_unref(avg);
```

**Comment by jcupitt** [comment]
This is good reading: https://github.com/libvips/ruby-vips/issues/155 I'd forgotten the details.

It seems just adding vips to your gemfile on macos might fix this.
